ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -0,0 +1,239 @@
+            } else {
+                actionListener.onFailure(buildOpenSearchStatusException(singleTensor));
+            }
+        } else {
+            // non batch API. This is to follow the previously code logic. Previously when making multiple requests to remote model,
+            // either one fails, we will return a failure response.
+            OpenSearchStatusException openSearchStatusException = null;
+            for (Map.Entry<Integer, ModelTensors> entry : sortedMap.entrySet()) {
+                if (entry.getValue().getStatusCode() < HttpStatus.SC_OK
+                    || entry.getValue().getStatusCode() > HttpStatus.SC_MULTIPLE_CHOICES) {
+                    openSearchStatusException = buildOpenSearchStatusException(entry.getValue());
+                    break;
+                }
+                modelTensors.add(entry.getKey(), entry.getValue());
+            }
+            if (openSearchStatusException != null) {
+                actionListener.onFailure(openSearchStatusException);
+            } else {
+                actionListener.onResponse(modelTensors);
+            }
+        }
+    }
+
+    private OpenSearchStatusException buildOpenSearchStatusException(ModelTensors modelTensors) {
+        try {
+            return new OpenSearchStatusException(
+                AccessController
+                    .doPrivileged(
+                        (PrivilegedExceptionAction<String>) () -> GSON.toJson(modelTensors.getMlModelTensors().get(0).getDataAsMap())
+                    ),
+                RestStatus.fromCode(modelTensors.getStatusCode())
+            );
+        } catch (PrivilegedActionException e) {
+            return new OpenSearchStatusException(e.getMessage(), RestStatus.fromCode(statusCode));
+        }
+    }
+
+    protected class MLResponseSubscriber implements Subscriber<ByteBuffer> {
+        private Subscription subscription;
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            this.subscription = s;
+            s.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onNext(ByteBuffer byteBuffer) {
+            responseBody.append(StandardCharsets.UTF_8.decode(byteBuffer));
+            subscription.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            log
+                .error(
+                    "Error on receiving response body from remote: {}",
+                    t instanceof NullPointerException ? "NullPointerException" : t.getMessage(),
+                    t
+                );
+            processResponse(statusCode, responseBody.toString(), parameters, tensorOutputs);
+            countDownLatch.getCountDownLatch().countDown();
+            response(tensorOutputs);
+        }
+
+        @Override
+        public void onComplete() {
+            processResponse(statusCode, responseBody.toString(), parameters, tensorOutputs);
```

Comment:
`processResponse` will return error for some failure case, for example if response body is empty. If these error happens, do we still need to run these code ?
```
            countDownLatch.getCountDownLatch().countDown();
            response(tensorOutputs);
```

line 225-227 duplicate with line 218-220 ? Tune these code and merge to process response method?

Reply:
Also find this problem in local env. When one response contains error, the listener will onFailure, and in `response` will onResponse Again, which cause unexpected error log. But I think we need to wait for all response and gather them together instead of stopping once we receive a error response.